File: src/main/java/net/codinux/kotlin/LambdasVsClosures.java
package net.codinux.kotlin;

import java.util.Optional;
import java.util.function.Supplier;

public class LambdasVsClosures {

    public void randomize(Supplier<Integer> numberSupplier) {
        // ...
    }

    public void canChangeVariables() {
        int i = 0;

        randomize(() -> {
            return ++i; // compile error, cannot change surrounding variable i
        });
    }

    public String canCallReturnInside(Optional<String> name) {
        // yeah, not the most senseful code, I know
        name.ifPresent(notNullName -> {
            return notNullName; // compile error, not possible in Java
        });

        return "Name not set";
    }

}

File: src/main/kotlin/net/codinux/kotlin/LambdasVsClosures.kt
package net.codinux.kotlin


class LambdasVsClosures {

  fun randomize(numberSupplier: () -> Int) {
    // ...
  }

  fun canChangeVariables() {
    var i = 0

    randomize {
      ++i // can change the value of surrounding variable i
    }
  }

  fun canCallReturnInside(name: String?): String {
    // yeah, not the most senseful code, I know
    name?.let {
      return it // not possible in Java
    }

    return "Name not set"
  }

}